src/charm.py

```diff
@@ -0,0 +1,176 @@
+        """
+        if not self.unit.is_leader():
+            return
+        if invalid_configs := self._invalid_configs():
+            self.unit.status = BlockedStatus(
+                f"The following configuration values are not valid: {invalid_configs}"
+            )
+            return
+        self._generate_root_certificate()
+        self.tls_certificates.revoke_all_certificates()
+        logger.info("Revoked all previously issued certificates.")
+        self.unit.status = ActiveStatus()
+
+    def _invalid_configs(self) -> list[str]:
+        """Returns list of invalid configurations.
+
+        Returns:
+            list: List of invalid config keys.
+        """
+        invalid_configs = []
+        if not self._config_ca_common_name:
+            invalid_configs.append("ca-common-name")
+        if not self._config_certificate_validity:
+            invalid_configs.append("certificate-validity")
+        return invalid_configs
+
+    def _on_certificate_creation_request(self, event: CertificateCreationRequestEvent) -> None:
+        """Handler for certificate requests.
+
+        Args:
+            event: CertificateCreationRequestEvent
+        """
+        if not self.unit.is_leader():
+            return
+        if invalid_configs := self._invalid_configs():
+            self.unit.status = BlockedStatus(
+                f"The following configuration values are not valid: {invalid_configs}"
+            )
+            event.defer()
+            return
+        if not self._root_certificate_is_stored:
+            self.unit.status = WaitingStatus("Root Certificates is not yet generated")
+            event.defer()
+            return
+        ca_certificate_secret = self.model.get_secret(label=CA_CERTIFICATES_SECRET_LABEL)
+        ca_certificate_secret_content = ca_certificate_secret.get_content()
+        certificate = generate_certificate(
+            ca=ca_certificate_secret_content["ca-certificate"].encode(),
+            ca_key=ca_certificate_secret_content["private-key"].encode(),
+            ca_key_password=ca_certificate_secret_content["private-key-password"].encode(),
+            csr=event.certificate_signing_request.encode(),
+            validity=self._config_certificate_validity,
+        ).decode()
+        self.tls_certificates.set_relation_certificate(
+            certificate_signing_request=event.certificate_signing_request,
+            certificate=certificate,
+            ca=ca_certificate_secret_content["ca-certificate"],
+            chain=[ca_certificate_secret_content["ca-certificate"], certificate],
+            relation_id=event.relation_id,
+        )
+        logger.info(f"Generated certificate for relation {event.relation_id}")
+
+
+def generate_password() -> str:
+    """Generates a random 12 character password.
+
+    Returns:
+        str: Password
+    """
+    chars = string.ascii_letters + string.digits
+    return "".join(secrets.choice(chars) for _ in range(12))
```

Comment:
This would be more secure and simpler, as there are more possiblities. We could also increase the size. Since this should never leave the charm and juju secret, I would go for at least 64.
```suggestion
    return secrets.token_bytes(12)
```